### PR TITLE
Ingress.ingressClassName field is readded if removed

### LINKED ISSUE #103546: Ingress.ingressClassName field is not readded if removed
#### What happened:
The default IngressClass was not applied to an Ingress when the ingressClassName field was deleted.

#### What you expected to happen:
The default IngressClass should be added to all new *or* updated Ingress' if they do not posses one.

#### How to reproduce it (as minimally and precisely as possible):
1. Create IngressClass, and add annotation to set as default IngressClass
2. Create Ingress without ingressClassName
3. Edit Ingress (using kubectl edit) and remove the ingressClassName field.
4. Get Ingress and note it's now missing an ingressClassName

#### Anything else we need to know?:
This was initially spotted due to using Lens to edit an Ingress which caused the ingressClassName to be added to the last-applied annotation, which a subsequent apply by the CI/CD system (the source file doesn't contain an ingressClassName) cause to get deleted.

As this is in AWS EKS we use the ALB Load Balancer component to manage an ALB based on the ingresses. The missing ingressClassName causes the ingress to be ignored by that component, and routing removed from the ALB.

Honestly I'm not sure enforcing ingressClassName rigorously is the best idea, but what does having an Ingress with no ingressClassName mean when there's a default IngressClass (esp. as the default is implemented by inserting that field in).

#### Environment:
- Kubernetes version (use `kubectl version`): v1.19.6-eks-49a6c0
- Cloud provider or hardware configuration: AWS EKS

### PR DESCRIPTION
**Pull request by mengjiao-liu**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
When the annotation `ingressclass.kubernetes.io/is-default-class` is set as the default IngressClass, if the field ingressClassName is removed, the value of the ingressClassName field will be re-added to the default ingressClass.

#### Which issue(s) this PR fixes:

Fixes #103546

#### Does this PR introduce a user-facing change?

```release-note
If the annotation `ingressclass.kubernetes.io/is-default-class` is set, and the ingressClassName field is removed when the ingress is updated, then the default ingress class is set as the value of the ingressClassName field.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: plugin/pkg/admission/network/defaultingressclass/admission_test.go
Comment (robscott) [inline plugin/pkg/admission/network/defaultingressclass/admission_test.go:272]: I may be missing something, but it looks like these fields are never actually set in the test cases. I want to be sure this change interacts well with the annotation, otherwise this PR LGTM.

**Review by robscott — COMMENTED** [review]
Thanks for the work on this @mengjiao-liu!

/triage accepted

**Comment by robscott** [comment]
One nit, I think this PR does need a release note as it is a change users will care about/appreciate. Labeling it as a bug fix may also be a bit more accurate and would increase the chances that we could cherry pick this fix back to earlier k8s versions. Thanks again for working on this!

File: plugin/pkg/admission/network/defaultingressclass/admission_test.go
Comment (mengjiao-liu) [reply plugin/pkg/admission/network/defaultingressclass/admission_test.go:272]: If I understand correctly, are you referring to this test case alone?
This is to test the situation when there is no default value, but it may not be necessary.

I am now modifying this test case as follows: there is a default ingress class before, and there is no default ingress class later, to test this situation.

**Review by jesse1478 — APPROVED** [review]

**Comment by mengjiao-liu** [comment]
> One nit, I think this PR does need a release note as it is a change users will care about/appreciate. 

Yes, it does require release notes. I have added the release notes, thanks for the reminder.


> Labeling it as a bug fix may also be a bit more accurate and would increase the chances that we could cherry pick this fix back to earlier k8s versions. Thanks again for working on this!

I read the discussion in #103546 issue. I think it was deliberately not applied the default ingress class when updating, so I think this is a feature. It’s okay to mark it as a bug, I will modify the label.

File: plugin/pkg/admission/network/defaultingressclass/admission_test.go
Comment (robscott) [reply plugin/pkg/admission/network/defaultingressclass/admission_test.go:272]: Sorry for the confusion, my concern was primarily that it didn't seem like `classAnnotationBefore` and `classAnnotationAfter` were being used by any test cases. It looks like `classAnnotationAfter` is used now, but I'm not seeing `classAnnotationBefore` specified anywhere.

**Comment by robscott** [comment]
>I read the discussion in #103546 issue. I think it was deliberately not applied the default ingress class when updating, so I think this is a feature. It’s okay to mark it as a bug, I will add the label.

That's fair, I don't have a strong opinion on this. I think it was deliberately left off, but also the resulting UX issue was not expected.

File: plugin/pkg/admission/network/defaultingressclass/admission_test.go
Comment (mengjiao-liu) [reply plugin/pkg/admission/network/defaultingressclass/admission_test.go:272]: This place call classAnnotationBefore:
https://github.com/kubernetes/kubernetes/pull/107018/files#diff-dbe2c160789adc3a41cf1f1157fe5733ecbf1c19fb15922e9747b719c1b18ca0R381

But the test cases do lack examples of using `classAnnotationBefore`, do you mean that?

Comment (robscott) [reply plugin/pkg/admission/network/defaultingressclass/admission_test.go:272]: Yep I noticed that there was code to implement it if it was set by any test cases, I just didn't see a non nil value for that field in any test cases.

Comment (mengjiao-liu) [reply plugin/pkg/admission/network/defaultingressclass/admission_test.go:272]: I know what you mean. This is my negligence. I accidentally ignored this situation while writing test cases. Thanks！

**Comment by dims** [comment]
looks like this is waiting on review from @deads2k adding Derek as well (listed in OWNERS file)

/assign @derekwaynecarr

**Comment by mengjiao-liu** [comment]
ping @derekwaynecarr @deads2k

File: plugin/pkg/admission/network/defaultingressclass/admission_test.go
Comment (robscott) [inline plugin/pkg/admission/network/defaultingressclass/admission_test.go:256]: From what I can tell this isn't actually used? Maybe just create a single "classes" field instead of "classesAfter"?

**Review by robscott — COMMENTED** [review]
Sorry I lost track of this @mengjiao-liu! This mostly LGTM.

**Pull request closed without merging**
